File: backend/app/schemas/student.py
"""Student schemas."""

from __future__ import annotations

from datetime import datetime
from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

FinancialStatusLiteral = Literal["low", "medium", "high"]
PlacementReadinessLiteral = Literal["low", "medium", "high"]
RiskLevelLiteral = Literal["low", "medium", "high"]


class StudentBase(BaseModel):
    roll_no: str = Field(min_length=1, max_length=40)
    name: str = Field(min_length=1, max_length=160)
    age: int = Field(ge=10, le=80)
    gender: str = Field(default="U", max_length=16)
    department_id: Optional[int] = None
    semester: int = Field(ge=1, le=12)
    attendance_pct: float = Field(ge=0.0, le=100.0)
    internal_marks: float = Field(ge=0.0, le=100.0)
    semester_marks: float = Field(ge=0.0, le=100.0)
    backlogs: int = Field(ge=0, le=50)
    fee_paid: bool = True
    fee_delay_days: int = Field(ge=0, le=730)
    financial_status: FinancialStatusLiteral = "medium"
    family_background: str = ""
    behavioral_indicators: str = ""
    extracurricular: str = ""
    placement_readiness: PlacementReadinessLiteral = "medium"
    counselor_remarks: str = ""
    faculty_notes: str = ""


class StudentCreate(StudentBase):
    pass


class StudentUpdate(BaseModel):
    name: Optional[str] = None
    age: Optional[int] = Field(default=None, ge=10, le=80)
    gender: Optional[str] = None
    department_id: Optional[int] = None
    semester: Optional[int] = Field(default=None, ge=1, le=12)
    attendance_pct: Optional[float] = Field(default=None, ge=0.0, le=100.0)
    internal_marks: Optional[float] = Field(default=None, ge=0.0, le=100.0)
    semester_marks: Optional[float] = Field(default=None, ge=0.0, le=100.0)
    backlogs: Optional[int] = Field(default=None, ge=0, le=50)
    fee_paid: Optional[bool] = None
    fee_delay_days: Optional[int] = Field(default=None, ge=0, le=730)
    financial_status: Optional[FinancialStatusLiteral] = None
    family_background: Optional[str] = None
    behavioral_indicators: Optional[str] = None
    extracurricular: Optional[str] = None
    placement_readiness: Optional[PlacementReadinessLiteral] = None
    counselor_remarks: Optional[str] = None
    faculty_notes: Optional[str] = None


class StudentOut(StudentBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    created_at: datetime
    updated_at: datetime
    department_code: Optional[str] = None
    department_name: Optional[str] = None
    latest_risk: Optional[RiskLevelLiteral] = None
    latest_confidence: Optional[float] = None


class StudentListQuery(BaseModel):
    q: Optional[str] = None
    department_id: Optional[int] = None
    risk: Optional[RiskLevelLiteral] = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=200)
    sort: str = Field(default="-created_at")


class TimelineEvent(BaseModel):
    type: Literal["prediction", "counseling", "note", "risk_snapshot"]
    timestamp: datetime
    title: str
    detail: dict


class StudentTimelineOut(BaseModel):
    student_id: int
    events: List[TimelineEvent]

File: frontend/src/types/student.ts
import type { RiskLevel } from '@/lib/constants';

export type FinancialStatus = 'low' | 'medium' | 'high';
export type PlacementReadiness = 'low' | 'medium' | 'high';

export interface Student {
  id: number;
  roll_no: string;
  name: string;
  age: number;
  gender: string;
  department_id: number | null;
  department_code: string | null;
  department_name: string | null;
  semester: number;
  attendance_pct: number;
  internal_marks: number;
  semester_marks: number;
  backlogs: number;
  fee_paid: boolean;
  fee_delay_days: number;
  financial_status: FinancialStatus;
  family_background: string;
  behavioral_indicators: string;
  extracurricular: string;
  placement_readiness: PlacementReadiness;
  counselor_remarks: string;
  faculty_notes: string;
  created_at: string;
  updated_at: string;
  latest_risk: RiskLevel | null;
  latest_confidence: number | null;
}

export interface StudentCreatePayload {
  roll_no: string;
  name: string;
  age: number;
  gender: string;
  department_id: number | null;
  semester: number;
  attendance_pct: number;
  internal_marks: number;
  semester_marks: number;
  backlogs: number;
  fee_paid: boolean;
  fee_delay_days: number;
  financial_status: FinancialStatus;
  family_background?: string;
  behavioral_indicators?: string;
  extracurricular?: string;
  placement_readiness: PlacementReadiness;
  counselor_remarks?: string;
  faculty_notes?: string;
}

export type StudentUpdatePayload = Partial<StudentCreatePayload>;

export interface TimelineEvent {
  type: 'prediction' | 'counseling' | 'note' | 'risk_snapshot';
  timestamp: string;
  title: string;
  detail: Record<string, unknown>;
}

export interface StudentTimeline {
  student_id: number;
  events: TimelineEvent[];
}
